api/src/main/java/com/epam/pipeline/manager/cloud/TemporaryCredentialsManagerImpl.java

```diff
@@ -56,32 +59,48 @@ public TemporaryCredentialsManagerImpl(final List<TemporaryCredentialsGenerator>
+                .filter(this::storageIdNotNull)
+                .map(DataStorageAction::getId)
+                .distinct()
+                .map(dataStorageManager::load)
+                .collect(Collectors.toList());
+        Assert.state(CollectionUtils.isNotEmpty(storages), "No storages were specified");
+        final TemporaryCredentialsGenerator credentialsGenerator = getCredentialsGenerator(storages);
 
-        final AbstractCloudRegion region = credentialsGenerator.getRegion(dataStorage);
+        final Map<Long, AbstractDataStorage> storagesById = storages.stream()
+                .collect(Collectors.toMap(AbstractDataStorage::getId, Function.identity()));
+        actions.forEach(action -> prepareAction(action, storagesById));
 
-        actions.forEach(action -> {
-            AbstractDataStorage loadedDataStorage =
-                    action.getId().equals(dataStorage.getId()) ? dataStorage : dataStorageManager.load(action.getId());
-            action.setBucketName(loadedDataStorage.getRoot());
-            action.setPath(loadedDataStorage.getPath());
-            AbstractCloudRegion loadedRegion = credentialsGenerator.getRegion(loadedDataStorage);
-            Assert.isTrue(Objects.equals(region.getId(), loadedRegion.getId()),
-                    "Actions shall be requested for buckets from the same region");
-        });
-
-        return credentialsGenerator.generate(actions, dataStorage);
+        return credentialsGenerator.generate(actions, storages);
     }
 
-    private TemporaryCredentialsGenerator getCredentialsGenerator(final AbstractDataStorage dataStorage) {
-        return Optional.ofNullable(MapUtils.emptyIfNull(credentialsGenerators).get(dataStorage.getType()))
+    private TemporaryCredentialsGenerator getCredentialsGenerator(final List<AbstractDataStorage> storages) {
+        final AbstractDataStorage storage = verifyAllTypesAreSameAngGetStorage(storages);
+        return Optional.ofNullable(MapUtils.emptyIfNull(credentialsGenerators).get(storage.getType()))
                 .orElseThrow(() -> new IllegalArgumentException(
                         messageHelper.getMessage(MessageConstants.ERROR_DATASTORAGE_NOT_SUPPORTED,
-                                dataStorage.getName(), dataStorage.getType())));
+                                storage.getName(), storage.getType())));
+    }
+
+    private void prepareAction(final DataStorageAction action, final Map<Long, AbstractDataStorage> storagesById) {
+        final AbstractDataStorage loadedDataStorage = storagesById.get(action.getId());
+        action.setBucketName(loadedDataStorage.getRoot());
+        action.setPath(loadedDataStorage.getPath());
+    }
+
+    private AbstractDataStorage verifyAllTypesAreSameAngGetStorage(final List<AbstractDataStorage> storages) {
+        Assert.state(storages.stream()
+                .map(AbstractDataStorage::getType)
+                .distinct()
+                .count() <= 1, "Storage types shall be the same");
```

Comment:
Could you please move all error messages (in `Assert` methods to message constants)

Reply:
done